tests/app/v2/notifications/test_get_notifications.py

```diff
@@ -339,6 +339,53 @@ def test_get_all_notifications_filter_by_single_status(api_client_request, sampl
     assert json_response["notifications"][0]["status"] == "pending"
 
 
+@pytest.mark.parametrize(
+    "internal_status, filter_status, should_get",
+    (
+        ("created", "accepted", True),
+        ("sending", "accepted", True),
+        pytest.param(
+            "delivered", "received", True, marks=pytest.mark.xfail(reason="bug in Notification.substitute_status impl")
+        ),
+        pytest.param(
+            "returned-letter",
+            "received",
+            True,
+            marks=pytest.mark.xfail(
+                reason=(
+                    "When we serialize a notification with status `returned-letter`, we display it as `received` "
+                    "instead. But we don't currently do the inverse when filtering on `received`. We probably should, "
+                    "but it would be a backwards-incompatible change and so may need flagging with API users before "
+                    "we fix this."
+                )
+            ),
+        ),
```

Comment:
Open question as to whether we just fix this, or want to notify API users as it would mean that anyone filtering on eg `notification_type=letter&status=received` is currently getting back only letters in the `delivered `status`, but would also get `returned-letter` if we "fixed" this.

Note also that we have an inconsistency around `returned-letter` now. We serialize notifications in that state as `received`, but on GET notifications they show up in filters for `status=failed` (but showing `status=received` in the response) 🙃

Reply:
I can see both sides of this issue. I however agree with @samuelhwilliams that the unique nature of a `returned-letter` as compared to undelivered emails or text messages, doesn't  wholly fit with `permanent-failure` and is probably best served by having a `returned` status.